Isolating core parameterization and architecting for recoupling

jatkinson1000 opened the pull request:
This PR should address and close #9 and close #10 and close #36 and close #40 and maybe close #35
- [X] Remove dependence on various SAM modules
- [X] Separate NN into separate functions (adressed in #23 and #28 which are merged here)
- [x] Clarify what is required as an input to the NN, and what needs to come from the model
- [x] Create a model specific wrapper function to feed into the NN - this can then be adapted for CAM/SCAM
- [x] Clarify physics processing of outputs from NN in `nn_convection_flux`

It also takes steps towards addressing #10 by adding a Makefile.
However, there should be discussion over the build process we wish to use.
The makefile currently requires a user to specify the location of NetCDF libraries to link against.

dorchard commented:
There are some big commented out blocks, I think due to the us dead-code removing code paths that won't be used because`rf_uses_rh` is `.false.`. Should we remove these in this PR totally then?

jatkinson1000 commented:
> There are some big commented out blocks, I think due to the us dead-code removing code paths that won't be used because`rf_uses_rh` is `.false.`. Should we remove these in this PR totally then?

I think the block can perhaps be removed from the SAM interface code, but am inclined to leave them in the parameterisation as it's one main block and functions at the end.
Perhaps something to ask Paul on Thursday about how serious they are to re-implement in future?
Other option would be to open a separate issue to remove `rf_uses_rh` code so that we can easily find the commit where it was present in future.

jatkinson1000 commented:
I have implemented the 3-layer structure, cleaned the code, and got rid of any implicit dependency on SAM.

These is still some physics to clarify, but this does not affect operation, so I'd like review on this now to merge and move to interfacing with SAM to check it still works OK.
This will also allow @raehik to look into CamFORT etc.

Moving from draft to review.

jatkinson1000 commented:
This PR will also close #10

raehik reviewed (approved):
LGTM. Reads nicely. Some changes to the Makefile to get things compiling on most easy platforms.

jatkinson1000 commented on NN_module/nn_cf_net.f90:
```diff
@@ -0,0 +1,289 @@
+        logits = matmul(z4, r_w5) + r_b5
+
+
+        ! Apply scaling and normalisation of output features
+        ! TODO: This is ugly, but not sure much can be done now.
+        i =1
+        logits(1:nrf) = (logits(1:nrf) * yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = nrf
+
+        i = i + 1
+        logits(out_dim_counter+1:out_dim_counter+nrf-1) = (logits(out_dim_counter+1:out_dim_counter+nrf-1)* yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = out_dim_counter + nrf-1
+
+        i = i + 1
+        logits(out_dim_counter+1:out_dim_counter+nrf-1) = (logits(out_dim_counter+1:out_dim_counter+nrf-1) * yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = out_dim_counter + nrf-1
+
+        i = i + 1
+        logits(out_dim_counter+1:out_dim_counter+nrf) = (logits(out_dim_counter+1:out_dim_counter+nrf) * yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = out_dim_counter + nrf
+
+        i = i + 1
+        logits(out_dim_counter+1:out_dim_counter+nrf) = (logits(out_dim_counter+1:out_dim_counter+nrf) * yscale_stnd(i))  +  yscale_mean(i)
+        out_dim_counter = out_dim_counter + nrf
+
```
Noice, will add.

jatkinson1000 replied:
Took a bit of rejigging of the above (including naughty use of `size` as a variable!) but in now and seems to still produce same numbers.

jatkinson1000 commented on NN_module/nn_convection_flux.f90:
```diff
@@ -0,0 +1,570 @@
+
+        ! These have been made module parameters
+        ! Hardcoded magic numbers
+        ! nrf is the fact results are supplied at lowest 30 half-model levels for sedimentation fluxes, and at 29 levels for fluxes
+        ! (as flux at bottom boundary is zero).
+        ! nrf = 30
+        ! nrfq = 29
```
Done

jatkinson1000 commented on NN_module/nn_convection_flux.f90:
```diff
@@ -0,0 +1,570 @@
+                !
+                do k=1, nrf
+                    ! TODO: Both these lines have dz*(1/dz) can we just remove this?
+                    precsfc(i,j) = precsfc(i,j) - q_tend_tot(k)*adz(k)*dz*rho(k)*rev_dz! removed the time step mult because q_tend_tot is already mult
```
Done.

jatkinson1000 commented on NN_module/nn_convection_flux.f90:
```diff
@@ -0,0 +1,570 @@
+
+!         real :: omn
+!             !! variable to store result of omegan function
+        real :: rev_dz
```
Don't think we can do it much earlier as `dz` is a function parameter so assignment needs to come after all other variable definitions? Leaving as is for now.

jatkinson1000 commented on NN_module/nn_convection_flux.f90:
```diff
@@ -0,0 +1,570 @@
+        real, intent(in) :: y_in(:)
+            !! Distance of column from equator (proxy for insolation and sfc albedo)
+        ! Fields from beginning of time step used as NN inputs
+        real, intent(in) :: t_i(:, :, :)
```
Yep, wanted to be very careful with this so just been through and done it for as much as I can. Have opened #40 to reflect this.

jatkinson1000 commented:
This will now close #40

jatkinson1000 commented:
Useful discussion with @paogorman who answered a number of questions I had around the science/numerics.
These will inform the refactor and as a result I will be making a few more (notable) changes on this PR so will likely want a re-review soon.
Summarised below:
* Questions:
  * `q` is total water, and `qn` is cloud water.  
    However, both are updated with adv + tot + sed, so what is the difference?  
    There are different checks on physical correctness (q checked at each addition, qn just checked after all added)?
    - `q` is total water content (vapour plus `qn`), `qn` is non-vapour water content.
    - We can probably comment out `qn` as it comes from SAM setup - in larger models it will be immediately precipitated out during a larger timestep.
    - In fact, just delete `qn`!!
  * gt lt misses 0.0 -> can we add?
    - Think about this, likely negligible effect from changing in reality
  * Check variable names and units
    - Seems OK
  * tendency is $\Delta\psi$?
    - Yes. Itr would be nicer if we can refactor for tendencies to actually be tendencies rather than changes
  * Flux is what related to `irhoadzdz`?
    - SAM equations are of the form:  
    $\text{tendency} = \text{d} h / \text{d}t = - 1 / \rho * \text{d} (\rho * u * h) / \text{d}x = - 1 / \rho * \text{d}(\text{flux}) / \text{d}x$
      So the `irhoadzdz` term comes from rearranging to get:
      $\text{d}h = - \text{irhoadzdz} * \text{d}({flux})$
  * What is `_tot`?
    - tot is a legacy from when the variable was total change. Now the tendencies are checked separately and this should really be called auto
    - CAM operates at the level of tendencies.
    - careful about removing dt as there is a dtn in irhodzdz

  - t is `energy / cp` beware of this, especially in the SAM-CAM conversion

jatkinson1000 commented on NN_module/nn_convection_flux.f90:
```diff
@@ -0,0 +1,570 @@
+                q(i,j,1:nrf) = q(i,j,1:nrf) + q_tend_tot(1:nrf)
+                t(i,j,1:nrf) = t(i,j,1:nrf) - q_tend_tot(1:nrf)*fac(1:nrf)
+
+                ! TODO ???
```
Sorted this - see comments in latest code.

jatkinson1000 commented:
This is a fairly sizeable PR with several linked issues at this point.
I propose we merge this back (after SAM integration check) and then open a subsequent PR linked to issues listed in the above comment refactoring further towards CAM interfacing by changing outputs to be tendencies rather than updated variables.
Thoughts @raehik @dorchard ?
